File: Arrays/04_ProductExceptSelf/solution.cpp
/*
Problem 4: Product of Array Except Self
Time Complexity: O(n), Space Complexity: O(1) excluding output
*/
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
              int n = nums.size();
        vector<int> result(n, 1);

        int prefix = 1;
        for (int i = 0; i < n; i++) {
            result[i] = prefix;
            prefix *= nums[i];
        }

        int suffix = 1;
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= suffix;
            suffix *= nums[i];
        }
        return result;
    }
};

int main() {
      Solution sol;
    vector<int> nums = {1,2,3,4};
    auto res = sol.productExceptSelf(nums);
    for (int x : res) cout << x << " ";  // 24 12 8 6
    return 0;
}

File: Arrays/04_ProductExceptSelf/solution.py
Arrays/04_ProductExceptSelf/solution.py"""
Problem 4: Product of Array Except Self
Given an integer array nums, return an array answer such that answer[i] 
is equal to the product of all the elements of nums except nums[i].
Time Complexity: O(n), Space Complexity: O(1) (excluding output array)
"""

from typing import List

def productExceptSelf(nums: List[int]) -> List[int]:
    n = len(nums)
    result = [1] * n

    # Left pass - calculate prefix products
    prefix = 1
    for i in range(n):
        result[i] = prefix
        prefix *= nums[i]

    # Right pass - multiply by suffix products
    suffix = 1
    for i in range(n - 1, -1, -1):
        result[i] *= suffix
        suffix *= nums[i]

    return result


if __name__ == "__main__":
    print(productExceptSelf([1,2,3,4]))  # [24,12,8,6]
    print(productExceptSelf([-1,1,0,-3,3]))  # [0,0,9,0,0]
